File: go/crypto.go
// Author: Tamaaxzcw
// GitHub: https://github.com/Tamaaxzcw

package main

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha512"
	"encoding/base64"
	"errors"
	"fmt"
	"io"
	"golang.org/x/crypto/pbkdf2"
)

const (
	saltSize   = 16
	ivSize     = 12
	tagSize    = 16
	keySize    = 32 // 256-bit
	iterations = 250000
)

func Encrypt(plainText, secret string) (string, error) {
	salt := make([]byte, saltSize)
	if _, err := io.ReadFull(rand.Reader, salt); err != nil {
		return "", err
	}

	key := pbkdf2.Key([]byte(secret), salt, iterations, keySize, sha512.New)
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	iv := make([]byte, gcm.NonceSize())
	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
		return "", err
	}
	
	// Seal akan menambahkan tag otentikasi secara otomatis
	cipherText := gcm.Seal(nil, iv, []byte(plainText), nil)

	// Gabungkan salt, iv, dan ciphertext
	payload := append(salt, iv...)
	payload = append(payload, cipherText...)

	return base64.StdEncoding.EncodeToString(payload), nil
}

func Decrypt(encryptedPayload, secret string) (string, error) {
	data, err := base64.StdEncoding.DecodeString(encryptedPayload)
	if err != nil {
		return "", err
	}

	if len(data) < saltSize+ivSize {
		return "", errors.New("encrypted data is too short")
	}

	salt := data[:saltSize]
	iv := data[saltSize : saltSize+ivSize]
	cipherText := data[saltSize+ivSize:]

	key := pbkdf2.Key([]byte(secret), salt, iterations, keySize, sha512.New)
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	plainText, err := gcm.Open(nil, iv, cipherText, nil)
	if err != nil {
		return "", errors.New("decryption failed: " + err.Error())
	}

	return string(plainText), nil
}

func main() {
	secretKey := "tamaaxcw-key"
	originalText := "Pesan ini dienkripsi menggunakan Go!"

	encrypted, err := Encrypt(originalText, secretKey)
	if err != nil {
		panic(err)
	}

	decrypted, err := Decrypt(encrypted, secretKey)
	if err != nil {
		panic(err)
	}

	fmt.Printf("Original: %s\n", originalText)
	fmt.Printf("Encrypted: %s\n", encrypted)
	fmt.Printf("Decrypted: %s\n", decrypted)
}

File: typescript/crypto.ts
// Author: Tamaaxzcw
// GitHub: https://github.com/Tamaaxzcw

import {
    createCipheriv, createDecipheriv, pbkdf2Sync, randomBytes,
} from 'crypto';

const ALGORITHM = 'aes-256-gcm';
const SALT_SIZE = 16;
const IV_SIZE = 12;
const TAG_SIZE = 16;
const KEY_DERIVATION_ALGO = 'sha512';
const ITERATIONS = 250000;
const KEY_SIZE = 32;

export function encrypt(text: string, secret: string): string {
    const salt = randomBytes(SALT_SIZE);
    const iv = randomBytes(IV_SIZE);
    const key = pbkdf2Sync(secret, salt, ITERATIONS, KEY_SIZE, KEY_DERIVATION_ALGO);

    const cipher = createCipheriv(ALGORITHM, key, iv);
    const encrypted = Buffer.concat([cipher.update(text, 'utf8'), cipher.final()]);
    const tag = cipher.getAuthTag();

    return Buffer.concat([salt, iv, tag, encrypted]).toString('base64');
}

export function decrypt(encryptedPayload: string, secret: string): string {
    const data = Buffer.from(encryptedPayload, 'base64');
    const salt = data.subarray(0, SALT_SIZE);
    const iv = data.subarray(SALT_SIZE, SALT_SIZE + IV_SIZE);
    const tag = data.subarray(SALT_SIZE + IV_SIZE, SALT_SIZE + IV_SIZE + TAG_SIZE);
    const encryptedText = data.subarray(SALT_SIZE + IV_SIZE + TAG_SIZE);

    const key = pbkdf2Sync(secret, salt, ITERATIONS, KEY_SIZE, KEY_DERIVATION_ALGO);

    const decipher = createDecipheriv(ALGORITHM, key, iv);
    decipher.setAuthTag(tag);

    const decrypted = Buffer.concat([decipher.update(encryptedText), decipher.final()]);
    return decrypted.toString('utf8');
}
